README.md

```diff
@@ -0,0 +1,492 @@
+
+```javascript
+_add('2', '2');
+// => '22'
+```
+
+`add`, on the other hand, throws if applied to arguments of the wrong types:
+
+```javascript
+add('2', '2');
+// ! TypeError: ‘add’ expected a value of type Number as its first argument; received "2"
+```
+
+Type checking is performed as arguments are provided (rather than once all
+arguments have been provided), so type errors are reported early:
+
+```javascript
+add('1');
+// ! TypeError: ‘add’ expected a value of type Number as its first argument; received "1"
+```
+
+### Types
+
+sanctuary-def defines `Type` values for JavaScript's built-in types:
+
+  * Nullary types:
+
+    - `$.Boolean`
+    - `$.Date`
+    - `$.Error`
+    - `$.Function`
+    - `$.Null`
+    - `$.Number`
+    - `$.RegExp`
+    - `$.String`
+    - `$.Undefined`
+
+  * Unary types:
+
+    - `$.Array`
+
+sanctuary-def also provides functions for defining types.
+
+#### `TypeVariable`
+
+Polymorphism is powerful. Not being able to define a function for all types
+would be very limiting indeed: one couldn't even define the identity function!
+
+```haskell
+TypeVariable :: String -> Type
+```
+
+Before defining a polymorphic function one must first define one or more type
+variables:
+
+```javascript
+const a = $.TypeVariable('a');
+const b = $.TypeVariable('b');
+
+//    id :: a -> a
+const id = def('id', {}, [a], x => x);
+
+id(42);
+// => 42
+
+id(null);
+// => null
+```
+
+The same type variable may be used multiple times:
+
+```javascript
+//    cmp :: a -> a -> Number
+const cmp = def('cmp', {}, [a, a], (x, y) => x < y ? -1 : x > y ? 1 : 0);
+
+cmp(42, 42);
+// => 0
+
+cmp('a', 'z');
+// => -1
+
+cmp('z', 'a');
+// => 1
+
+cmp(0, '1');
+// ! TypeError: ‘cmp’ expected a value of type Number as its second argument; received "1"
+```
+
+#### `NullaryType`
+
+`NullaryType` is used to construct types with no type variables. The `Number`
+type exported by sanctuary-def is defined via `NullaryType`, as are many of the
+other exported types.
+
+To define a nullary type one must specify:
+
+  - the type's name;
+  - a predicate which accepts any JavaScript value and returns `true` if
+    (and only if) the value is a member of the type; and
+  - a thunk which returns an arbitrary value of the type.
+
+```haskell
+NullaryType :: String -> (* -> Boolean) -> (-> t) -> Type
+```
+
+For example:
+
+```javascript
+//    Integer :: Type
+const Integer = $.NullaryType(
+  'sanctuary-def/Integer',
+  x => Object.prototype.toString.call(x) === '[object Number]' &&
+       Math.floor(x) === Number(x) &&
+       x >= Number.MIN_SAFE_INTEGER &&
+       x <= Number.MAX_SAFE_INTEGER,
+  () => 0
+);
+
+//    rem :: Integer -> Integer -> Integer
+const rem = def('rem', {}, [Integer, Integer], (x, y) => {
+  if (y === 0) throw new Error('Cannot divide by zero');
+  return x % y;
+});
+
+rem(42, 5);
+// => 2
+
+rem(42, 0.5);
+// ! TypeError: ‘rem’ expected a value of type Integer as its second argument; received 0.5
+```
```

Comment:
This example is great, you could go on to show how to use `NullaryType` to define a non-zero integer, and take that as the second argument. Thus you needn't check for it in the function!